### PR TITLE
feat: init wizard v2 

### PR DESCRIPTION
**Pull request by aorumbayev**
Implements https://github.com/orgs/algorandfoundation/projects/1?pane=issue&itemId=53079977

## Proposed Changes

This is a draft moving the PoC for init wizard v2 discussed in https://github.com/algorandfoundation/algokit-cli/pull/411. 
The adr is split into 2 parts and this PR will be covering the improvements in wizard v2 + changes in related template repos. 

Currently the draft contains poc code shown during the init wizard v2 discussions for the ADR. The PR is to be refined and turned into a full implementation by the time it leaves the draft state. 

To interact with the PoC simply checkout the branch and access algokit inside poetry environment. Please note that `tealscript` template variation is not yet fully covered as poc was demoing a happy path from start to fullstack with puya + react components (this is to be patched soon). 

## Template PRs 

- FullStack -> https://github.com/algorandfoundation/algokit-fullstack-template/pull/19
- React -> https://github.com/algorandfoundation/algokit-react-frontend-template/pull/25
- Beaker -> https://github.com/algorandfoundation/algokit-beaker-default-template/pull/59
- Puya -> https://github.com/algorandfoundation/algokit-puya-template/pull/5
- new base template is available at -> https://github.com/algorandfoundation/algokit-base-template/pull/1

**Above has to be merged first** -> branch refs has to be then removed on this PR, afterwards ready to release new cli version

### CODE REVIEW

File: src/algokit/cli/init.py [hunk 1]
```diff
@@ -61,14 +81,17 @@ def _get_blessed_templates() -> dict[str, BlessedTemplateSource]:
         "puya": BlessedTemplateSource(
             url="gh:algorandfoundation/algokit-puya-template",
             description="Official starter template for Puya applications (Dev Preview, not recommended for production)",
+            branch="poc/wizard_v2",
```
Comment (aorumbayev) [inline src/algokit/cli/init.py:84]: @robdmoore @neilcampbell please note this is a temporary change done for the sake of adr given that dependant templates also rely on changes on specific branch

File: src/algokit/cli/init.py [hunk 2]
```diff
@@ -85,10 +108,11 @@ def _get_blessed_templates() -> dict[str, BlessedTemplateSource]:
 
 
 def validate_dir_name(context: click.Context, param: click.Parameter, value: str | None) -> str | None:
-    if value is not None and not re.match(r"^[\w\-.]+$", value):
+    if value is not None and not re.match(r"^[\w\-.\\/]+(?:[\w\-. ]+)*$", value):
```
Comment (aorumbayev) [inline src/algokit/cli/init.py:111]: these are to also allow specifying a target folder as a path - currently it always assumes that if its a folder in local fs it has to be relative to root. This change allows setting this param as 'path/to/some/nested/dir' - this way can also ensure that when fullstack is doing init on child templates instead of doing `mkdir` checks to drop them into a workspace structure it can specify the nested path directly in the name parameter

File: src/algokit/core/bootstrap.py [hunk 3]
```diff
@@ -34,15 +35,17 @@ def bootstrap_any(project_dir: Path, *, ci_mode: bool) -> None:
         bootstrap_npm(project_dir)
 
 
-def bootstrap_any_including_subdirs(base_path: Path, *, ci_mode: bool) -> None:
+def bootstrap_any_including_subdirs(base_path: Path, *, ci_mode: bool, depth: int = 0) -> None:
```
Comment (aorumbayev) [inline src/algokit/core/bootstrap.py:38]: @robdmoore @neilcampbell these are to ensure that we dive at least 2 folders deep by default instead of 1 level that we get right now since new template structure is by default require 2 nesting levels 'src/{some_project}` and without these changes bootstrap won't get to them. We can also explore parametrizing lookup depth as a parameter you can pass to bootstrap

**Comment by aorumbayev** [comment]
@robdmoore @neilcampbell still keeping the pr in draft for ~1 more day until related prs on templates are opened too, added extra tests here and minor tweaks but should mostly be ready for review already

**Comment by aorumbayev** [comment]
@neilcampbell @robdmoore added extra logic as per latest discussions, adds optional behaviour that transforms any standalone template into a workspace structure by default OR if its already init'd from root of a workspace project it will auto inject the standalone template under expected folder. Let me know what you think, ignore the tests for now - will patch on monday

**Comment by aorumbayev** [comment]
@robdmoore @neilcampbell added extra documentation and usage of https://github.com/algorandfoundation/algokit-base-template to enforce workspace structure by default. This base is reused within all official templates (including fullstack) and decouple files that we want to see in the base structure from individual templates. 

The best way to get a hang of the latest state is to checkout the branch and interact with the new wizard within poetry shell. 

Would appreciate initial round of reviews as there are still certain open questions: 

1. Do we want to enable bootstrap command to also include `depth` which will allow users to control the lookup depth? Right now im defining a parameter on init command that controls initial bootstrap depth because now with default workspace structure we need that bootstrap invocation as part of init to go down 2 levels instead of 1. However haven't added it to bootstrap commands separately - what's your opinions on this so far? I think deffo makes sense to do 2 level lookup on bootstrap within init flow as well as adding this to main bootstrap group (however i wonder if this can be considered out of scope for this pr)
2. Took some time to patch existing init tests, and they all were mostly tailored for the old wizard so im adding the --no-workspace everywhere in the old tests, there is a few new tests specifically for the init wizard v2 flow, will add a few extra edge cases but if youll spot anything critical in the pr that you think is not covered with tests let me know. 
3. PRs are still pending on beaker, react, puya and fullstack (you should be able to interact with the wip changes when checking out and running algokit locally from this branch those). Want to ensure everything is solid on cli side first before i jump to finalizing the template changes (luckily should be fairly small and won't have anything backwards incompatible). Also let me know what you think of the usage of https://github.com/algorandfoundation/algokit-base-template, i think it simplifies the injection of common workspace structure by a good margin. 


cc @PhearZero @joe-p @CiottiGiorgio @iskysun96 - feel free to try it out as well if you want to provide preliminary feedback, please note the branch might still have further commits AFTER we coordinate opening prs on templates. Template prs will be merged first then we will make the prerelease on the cli and will coordinate with @Loedn to have an interactive feedback session to patch any last minute tweaks before end of month. Otherwise if all good i'll jump to part 2 (which will add orchestration on top of whats enforced with the new features in this pr)

File: docs/features/init.md [hunk 4]
```diff
@@ -4,45 +4,54 @@ The `algokit init` [command](../cli/index.md#init) is used to quickly initialize
...
-   Ruff
-🎤 Do you want to use a Python formatter (via Black)?
-   Yes
-🎤 Do you want to use a Python type checker (via mypy)?
-   Yes
-🎤 Do you want to include unit tests (via pytest)?
-   Yes
-🎤 Do you want to include Python dependency vulnerability scanning (via pip-audit)?
-   Yes
+? How would you like to build your project? Smart Contract
```
Comment (neilcampbell) [inline docs/features/init.md:19]: Docs here will need updating based on the changes we spoke about.

File: docs/features/init.md [hunk 5]
```diff
@@ -4,45 +4,54 @@ The `algokit init` [command](../cli/index.md#init) is used to quickly initialize
...
+? Which language would you like to use for the smart contract? Python
+? Would you like to include a frontend component? Yes
+? Name of project / directory to create the project in:  my-cool-app
 ```
 
+Once above 4 questions are answered, the `cli` will start instantiating the project and will start asking questions specific to the template you are instantiating. By default official templates such as `puya`, `fullstack`, `react`, `beaker` include a notion of a `preset`. If you want to skip all questions and let the tool preset the answers tailored for a starter project you can pick `Starter`, for a more advanced project that includes unit tests, CI automation and other advanced features pick `Production`. Lastly, if you are a prefer to modify the experience and tailor the template to your needs pick `Custom` preset.
```
Comment (neilcampbell) [inline docs/features/init.md:25]: Docs here will need updating based on the changes we spoke about.

File: src/algokit/cli/init.py [hunk 6]
```diff
@@ -29,12 +36,57 @@
...
+    """
+
+    SMART_CONTRACT = auto()
+    DAPP_FRONTEND = auto()
+    CUSTOM_TEMPLATE = auto()
+
+
+class ContractLanguage(Enum):
+    """
+    For general purpose languages that have corresponding smart contract languages
```
Comment (neilcampbell) [inline src/algokit/cli/init.py:51]: This comment needs tweaking to factor in the added pyteal option.

File: tests/portability/test_pyinstaller_binary.py [hunk 7]
```diff
@@ -42,11 +42,16 @@ def test_algokit_init(
     cli_path: str,
     dummy_algokit_template_with_python_task: dict[str, Path],
 ) -> None:
+    # TODO: revisit to improve
```
Comment (neilcampbell) [inline tests/portability/test_pyinstaller_binary.py:45]: To answer the question of adding to the PATH, you can do this in the GitHub action.
```
echo "$TEST_CLI_PATH" >> $GITHUB_PATH
```

File: src/algokit/cli/init.py [hunk 8]
```diff
@@ -500,3 +659,69 @@ def git(*args: str, bad_exit_warn_message: str) -> bool:
...
+        and config.get("project", {}).get("type") == ProjectType.WORKSPACE.value
+        and is_standalone
+        and use_workspace
+    ):
+        sub_projects_path = config.get("project", {}).get("projects_root_path") or "projects"
+        projects_root = cwd / sub_projects_path
+        logger.debug(f"Workspace structure detected! Moving the project to be instantiated into {projects_root}")
+        return projects_root / project_path.name
+
+    return project_path
```
Comment (neilcampbell) [inline src/algokit/cli/init.py:704]: If it's a "standalone project being instantiated inside an existing workspace project and use_workspace is False", it appears would result in a weird project, where the standalone template is applied over top of the workspace? Would it be better to bail out with a message that it's not a supported configuration or simply ignore the `use_workspace=False`

File: docs/cli/index.md [hunk 9]
```diff
@@ -641,9 +648,17 @@ Whether to run algokit bootstrap to install and configure the new project's depe
...
+### --workspace, --no-workspace
+Whether to prefer structuring standalone projects as part of a workspace.
+
+
 ### -a, --answer <key> <value>
 Answers key/value pairs to pass to the template.
 
+
+### --bootstrap-depth <bootstrap_depth>
+Maximum depth to apply when running bootstrap in new project. Defaults to 2 levels down in the folder tree with --workspace and 1 when used with --no-workspace.
```
[the remark this answers is not in the document]
Comment (neilcampbell) [reply docs/cli/index.md:660]: I guess that's my answer to `1. Do we want to include an optional arg to control max bootstrap lookup depth?`

**Comment by robdmoore** [comment]
Got an error on Windows, also it didn't ask me for smart contract language
image

**Comment by aorumbayev** [comment]
@neilcampbell all templates should be up to date with latest - so running algokit cli from this branch should not have any e2e problems (only tested on windows so far though, but binaries portability tests add a bit more confidence). 

A PR for fullstack is available here https://github.com/algorandfoundation/algokit-fullstack-template/pull/19, once base template is public i'll open prs on remaining repos (no reason to open now since all will fail due to base being private repo)

File: tests/portability/test_pyinstaller_binary.py (hunk 7 above)
Comment (aorumbayev) [reply tests/portability/test_pyinstaller_binary.py:45]: To be addressed separately, explicit notes are added in the codebase

File: docs/cli/index.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (aorumbayev) [reply docs/cli/index.md:660]: Agree, removing this from being exposed, lets revisit if there are concrete feature requests

**Comment by aorumbayev** [comment]
@robdmoore thanks for reporting on the bugs so far - they are fixed now, let me know if there are any further comments. Also refer to updated pr description that includes refs to prs on respective react, fullstack, beaker and puya templates

File: src/algokit/cli/init.py (hunk 2 above)
Comment (aorumbayev) [reply src/algokit/cli/init.py]: this is now reverted back to original implementation, a different method takes care of moving the project to expected projects folder inside the workspace without needed to modify the existing init command semantics

File: docs/cli/index.md [hunk 10]
```diff
@@ -604,12 +611,12 @@ Name of the project / directory / repository to create.
...
 
 ### -t, --template <template_name>
-Name of an official template to use. To see a list of descriptions, run this command with no arguments.
+Name of an official template to use. To choose interactively, run this command with no arguments.
 
 
 * **Options**
 
-    beaker | tealscript | puya | react | fullstack | playground
+    TemplateKey.TEALSCRIPT | TemplateKey.PUYA | TemplateKey.REACT | TemplateKey.FULLSTACK | TemplateKey.BEAKER | TemplateKey.BASE | MiscTemplateKey.PLAYGROUND
```
Comment (neilcampbell) [inline docs/cli/index.md:619]: Is there any way to expand the values when generating the docs?

File: docs/features/init.md [hunk 11]
```diff
@@ -4,45 +4,52 @@ The `algokit init` [command](../cli/index.md#init) is used to quickly initialize
...
-   Yes
-🎤 Do you want to include unit tests (via pytest)?
-   Yes
-🎤 Do you want to include Python dependency vulnerability scanning (via pip-audit)?
-   Yes
+? Which of these options best describes the project you want to start? `Smart Contract` | `Dapp Frontend` | `Smart Contract & Dapp Frontend` | `Custom`
+? Name of project / directory to create the project in:  my-cool-app
 ```
 
+Once above 2 questions are answered, the `cli` will start instantiating the project and will start asking questions specific to the template you are instantiating. By default official templates such as `puya`, `fullstack`, `react`, `beaker` include a notion of a `preset`. If you want to skip all questions and let the tool preset the answers tailored for a starter project you can pick `Starter`, for a more advanced project that includes unit tests, CI automation and other advanced features pick `Production`. Lastly, if you are a prefer to modify the experience and tailor the template to your needs pick `Custom` preset.
```
Comment (neilcampbell) [inline docs/features/init.md:23]:
```suggestion
Once above 2 questions are answered, the `cli` will start instantiating the project and will start asking questions specific to the template you are instantiating. By default official templates such as `puya`, `fullstack`, `react`, `beaker` include a notion of a `preset`. If you want to skip all questions and let the tool preset the answers tailored for a starter project you can pick `Starter`, for a more advanced project that includes unit tests, CI automation and other advanced features, pick `Production`. Lastly, if you prefer to modify the experience and tailor the template to your needs, pick the `Custom` preset.
```

**Review by neilcampbell — APPROVED** [review]

File: docs/cli/index.md (hunk 10 above)
Comment (aorumbayev) [reply docs/cli/index.md]: Done

**Comment by aorumbayev** [comment]
@joe-p @PhearZero just fyi, as agreed i updated child templates to come shipped with type `contract` by default in the .toml. So now we got `workspace`, `backend`, `frontend`, `contract` types (no templates containing `backend` yet but we can assume its a supported category). As for suggestion to make opening in code workspaces optional (for full stack), i forgot to mention that cli already supports a flag `--no-ide` which you can pass to instruct it to ignore auto opening the IDE @joe-p , remaining bits are going to be revisited during orchestration work from next week and we may have another feedback session some time next week to go over the orchestration as well.
